gui: Add option to hide file extensions

me25542 opened the pull request:
Resolves #4426.

Adds "Show File Extensions" toggle to user interface menu, directly below "Sort Files".
- If this toggle is enabled, behavior is unchanged.
- If this toggle is disabled, file extensions are hid from the file browser, print preview, and printing screens.

Increases RAM usage slightly; at least ~104 extra bytes could be used at any given time.

Tested and seems to work on a MK4S.

BFW-6704

CZDanol committed:
FilamentSensorADC: Rework to use hysteresis

Rework the ADC FS logic to get rid of the fixed FS span.

BFW-5338

CZDanol committed:
Remove MI_X_FS_SPAN

Not using fs_span, get rid of it

BFW-5338

CZDanol committed:
Remove fs_span from MI_X_CALIBRATION_FROM_USB

BFW-5338

CZDanol committed:
Remove fsensor span from config_store

BFW-5338

CZDanol committed:
Rename MI_FS_REF -> MI_FS_REF_NINS

I will be adding REF_INS in the next commit

BFW-5338

CZDanol committed:
ScreenToolheadSettings: Add MI_FS_REF_INS

BFW-5338

danopernis committed:
phstep: Home imprecisely for calibration also in GUI

BFW-7465
BFW-7547

CZDanol committed:
Move extruder_fs_value_span out of config_store defaults

It is no longer relevant to the config store.

The CalibrateInserted is now a bit mesy, but I intend to rework the whole calibration in the upcoming PR, so I would keep it this way for now.

BFW-5338

michalrudolf committed:
Add preemptive move of Y to avoid ventilation grille mechanism

BFW-8021

danopernis committed:
Update translations

vorner committed:
Selftest: Allow ignoring noise of loadcell

And proceed with the test anyway.

BFW-7542.

bkerler committed:
Allow nozzle diameter of 0.1mm and 0.15mm

danopernis committed:
Use only katakana in japanese translations

danopernis committed:
Fix unittest segfault

vorner committed:
auto_retract: Don't do eStall detection during it

As these happen fast & outside of the print, the motor may skip, but
that's OK. Don't report as filament stuck (because likely it isn't).

BFW-7545

danopernis committed:
Regenerate fonts

danopernis committed:
Fix fuzzy ukrainian translations

danopernis committed:
git subrepo pull --force lib/Prusa-Error-Codes

subrepo:
  subdir:   "lib/Prusa-Error-Codes"
  merged:   "d64eb1056d"
upstream:
  origin:   "git@github.com:prusa3d/Prusa-Error-Codes.git"
  branch:   "master"
  commit:   "d64eb1056d"
git-subrepo:
  version:  "0.4.9"
  origin:   "???"
  commit:   "???"

o-vesely committed:
Fix enqueuing G12 only on nozzle cleaning printers

BFW-7598

CZDanol committed:
HACK: Make the new fsensor_adc work with ye old selftest

The selftest expects the fsensor to report HasFilament/NoFilament right after nins calibration is done, even though ins calibration is not yet done.

This forced me to reintroduce filament sensor span just for that situation.

Temporary solution till the new selftest comes.

BFW-5338

schrodlm committed:
Integrate vent grille control into a gcode (M870)

BFW-7395
BFW-8021

michalrudolf committed:
Belt tuning: reset LEDs after belt tuning exits

BFW-7413

michalrudolf committed:
Belt tuning: move static variables to nameless namespace

BFW-7413

michalrudolf committed:
Belt tuning: Add finish screen

BFW-7413

michalrudolf committed:
Belt tuning: Add number next to the knob when tuning belt tension

BFW-7413

michalrudolf committed:
Belt tuning: remove unreachable break

BFW-7413

michalrudolf committed:
Belt tuning: Pass multiplied PhaseData to accomodate 0.5 step (adjusting frequency)

BFW-7413

CZDanol committed:
Cherry-pick fixups

michalrudolf committed:
Loadcell selftest: Add QR on tap screen when loadcell data are too noisy

BFW-7542

danopernis committed:
git subrepo pull --force lib/Prusa-Error-Codes

subrepo:
  subdir:   "lib/Prusa-Error-Codes"
  merged:   "a0138327b4"
upstream:
  origin:   "git@github.com:prusa3d/Prusa-Error-Codes.git"
  branch:   "master"
  commit:   "a0138327b4"
git-subrepo:
  version:  "0.4.9"
  origin:   "???"
  commit:   "???"

schrodlm committed:
Implement automatic chamber vents control

Cherry-pick with added C1 values
Co-Authored-By: Dano Pernis <danopernis@gmail.com>
Co-Authored-By: Ondřej Veselý <o.vesely42@gmail.com>

BFW-7395
BFW-8021

schrodlm committed:
Integrate automatic vent control into printer logic

BFW-7395
BFW-8021

schrodlm committed:
Add GUI for automatic vent control

BFW-7395
BFW-8021

danopernis committed:
Kick the dog while sleeping in factory

BFW-7627

michalrudolf committed:
Belt tuning: Remove Recheck screens

BFW-7413

michalrudolf committed:
Belt tuning: Adjust wizard process

Add alignment_issue screen
Adjust finish screen
Edit texts
disable motors when appropriate

BFW-7413

michalrudolf committed:
Belt tuning: Edit show_tensions screen's text

BFW-7413

michalrudolf committed:
Rename belt tuning screen to avoid confusing with old belt tuning

BFW-7413

michalrudolf committed:
Belt tuning: Fix buddy unreachable

BFW-7413

Commit:
MMU: Obey auto-retracted in MMU-unload

Commit:
MMU: unify unloading in Unload and LoadToNozzle

CZDanol committed:
Revert "probe_analysis: Use fixed point for samples"

This reverts commit 55eb86e2d04b62041a123b2a2c0549309e2c7516.

michalrudolf committed:
Belt tuning: calculation adjustments

Optimal belt frequency depends on which belt has higher frequency (interchangable optimal values)
Consider valid if frequencies are in valid range (90 - 98 Hz)

BFW-7413

CZDanol committed:
loadcell: Make do without filter settling

If you initialize the filters to the first provided value,
they do not require the settling period.
This shaves off 400 ms (the settling period) from each probe,
especially handy during MBL :)

BFW-7649 BFW-7672

vorner committed:
Revert "Websocket release"

This reverts commit a6312dd7e078d60d52a3867800c4ac6685049770.

CZDanol committed:
Reapply "Temporarily disable M600 upon E-stall detected"

This reverts commit 72fc2ce0288c4d297c9374daedd77d87567bf6ac.

o-vesely committed:
Filament unload wait only for ExtruderFS

BFW-7435

michalrudolf committed:
Update PO files

michalrudolf committed:
Update POT file

CZDanol committed:
loadcell: Further increase ANALYSIS_WINDOW_SIZE

Still encoutenered a printer that wasn't able to fit the data to the buffer.

BFW-7649

o-vesely committed:
Update minimum resonance to 60 Hz to have it epi-safe

o-vesely committed:
Add no_parser to M109

BFW-7708

o-vesely committed:
Add "Purge" as a response

BFW-7708

o-vesely committed:
Add "purging" as a print status message

BFW-7708

o-vesely committed:
Nozzle cleaning failed with retry purging in a FSM

BFW-7708

o-vesely committed:
Enable Auto-Retract on XL

BFW-7708

michalrudolf committed:
Auto retract: Replace custom EStallDisabler with common EStallBlocker

michalrudolf committed:
Retraction tracking: Add filament retraction before nozzle cleaning

BFW-7621

vorner committed:
Reapply "Websocket release"

This reverts commit dd70eadaf46be63042462d5cb2ea6d2b8a1be0dc.

o-vesely committed:
Constexpr instead of consteval in TemplateString

michalrudolf committed:
Auto Retract: Use retracted distances instead of bitset

Replace filament_auto_retracted_bitset with filament_retracted_distances in persistent storage
Adjust AutoRetract class to use float retracted distances on each nozzle and handle retraction based on them
Also introduce invalid value that signals unknow state
Add invalid_retracted_distance to config_store_ns::

BFW-7621

michalrudolf committed:
Auto Retract: Use retracted distances instead of bitset

Replace filament_auto_retracted_bitset with filament_retracted_distances in persistent storage
Adjust AutoRetract class to use float retracted distances on each nozzle and handle retraction based on them
Also introduce invalid value that signals unknow state
Add invalid_retracted_distance to config_store_ns::

BFW-7621

michalrudolf committed:
Auto retract: Add migration from bitset to array of retracted distances

BFW-7621

michalrudolf committed:
Auto retract: Add migration from bitset to array of retracted distances

BFW-7621

o-vesely committed:
Add utility for concatenating two std::array objects

BFW-7713

o-vesely committed:
Autolayout: Use std::span

Use span instead of initializer list

BFW-7713

o-vesely committed:
Add footer support to GUI frames

Introduces functionality for GUI frames to include an optional footer.

BFW-7713

vorner committed:
footer: Show dashes for unknown XL enclosure temp

Fix the footer item, show dashes instead of 0 when the temperature is
unknown.

BFW-7717.

vorner committed:
XL enclosure temp timer fix

The marlin's print_duration is actually in seconds, not milliseconds -
the comment there was wrong.

BTW-7717.

o-vesely committed:
Expand with_footer to allow more frames

This way WithFooter couldnt take custom frames that were hidden in anon-namespaces. This way this takes any frame which still has parent as first argument.

BFW-7708

michalrudolf committed:
Remove deprecated function in nozzle_cleaning_failed scenario

BFW-7621
BFW-7708

o-vesely committed:
Expand autoretract to allow custom progress tracking

Added a callback argument for custom progress tracking

BFW-7708

michalrudolf committed:
Retract tracker: Introduce feature for tracking retracted distance on each hotend (during printing)

BFW-7621

michalrudolf committed:
Retract tracker: Add overwriting EEPROM values on the end of the print

BFW-7621

michalrudolf committed:
Retract tracker: Update retracted distances based on planner E moves

BFW-7621

CZDanol committed:
screen_filament_detail:MI_TOGGLE: Add support for inverted values

BFW-7696

CZDanol committed:
FilamentTypeParameters: is_flexible -> do_not_auto_retract

BFW-7696

o-vesely committed:
Add no_parser to M109

BFW-7708

schrodlm committed:
Remove unused includes in first layer selftest

michalrudolf committed:
Revert "Auto Retract: Use retracted distances instead of bitset"

This reverts commit 920399b6c9d60d3fee58076c29adb3e5bce6814d.

michalrudolf committed:
Revert "Auto retract: Add migration from bitset to array of retracted distances"

This reverts commit 8d44794cfebf710a491f2b118814b27efff46821.

CZDanol committed:
Revert "Nozzle cleaning failed with retry purging in a FSM"

This reverts commit dad04c7b4803ee97c3d7ae3e357f0a89d7fc2e03.

CZDanol committed:
Revert "Add "purging" as a print status message"

This reverts commit 56e0c8ef49fb0cc3df00b3be54af96a8a2075594.

CZDanol committed:
Revert "Add "Purge" as a response"

This reverts commit 23ae830f8748f517d676c4aa0f083735d81a9107.

CZDanol committed:
Revert "Add no_parser to M109"

This reverts commit e078dc55f4972a932d89790bb13733062b920df3.

o-vesely committed:
Nozzle cleaning failed with retry purging in a wizard FSM

BFW-7708

o-vesely committed:
Remove printer type link prefix from FrameQRPrompt

BFW-7708

Commit:
Make sure we never park below max_printed_z. Probably solves BFW-7734, but no reproducer available yet

o-vesely committed:
Add special point for nozzle cleaning failed parking

This is needed for having enough room under nozzle to remove filament if needed.

BFW-7708

CZDanol committed:
auto_retract: Make disablable

BFW-7721

michalrudolf committed:
Update sensor_data even if metrics are disabled

There are also used in SensorInfo screen and some other places

BFW-7706

CZDanol committed:
auto_retract: Add menu item for disabling

BFW-7721

CZDanol committed:
auto_retract: De-duplicate functions

Instead of providing overloads for the functions, provide a defaulted argument.

I wanted to avoid having to include marlin_vars into the header, so I introduced a local current_hotend instead.

BFW-7721

CZDanol committed:
Raise above max_printed_z on print end

We don't want to crash into sequential prints

BFW-7763 BFW-6609

michalrudolf committed:
Add atomic wrapper with templated memory order

BFW-7706

michalrudolf committed:
Rewrite sensor_data variables to new Atomic

BFW-7706

michalrudolf committed:
Remove duplicate RateLimiter class from app_metrics

BFW-7706

michalrudolf committed:
Add sandwichTemp and splitterTemp to sensor_data and update them outside of metric handling

BFW-7706

schrodlm committed:
Move G26 impl to a FirstLayer class

BFW-7738

schrodlm committed:
Fix first layer selftest

BFW-7738

Commit 3897cd7 broke a self-test by fixing a bug that allowed the test to run in a hacky, parallel way. This test now correctly runs sequentially, causing a failure where stateMblFinished expects a FirstLayer instance that was already destroyed in the previous step. The fix updates the test logic to handle the new, sequential behavior.

schrodlm committed:
Improve preheat in first layer selftest

Partially BFW-7738 (cleaning up)

schrodlm committed:
Effort to inline remaining gcodes in first layer selftest

Partially BFW-7738 (cleaning up)

schrodlm committed:
Remove unused stales of first layer selftest

Partially BFW-7738 (cleaning up)

schrodlm committed:
Disable the live_z GUI when not usable

Partially BFW-7738 (cleaning up)

This is to improve UX, when I opened the selftest for the first time, the UX makes it look like you should be able to manipulate the Z height right away which is not true.

BTW I know there are probably better way to disable it but I spent way too long on this detail already, if the reviewer has a cleaner approach please let me know in the review.

CZDanol committed:
Fix mapi::extruder_move double-applying modifiers

BFW-7770

Commit:
MMU: fixup unittests

vorner committed:
XL: Don't crash tools because of derectract

*This is a temporary workaround only*.

For some reason, the auto-derectract causes the purge line to lose 5 mm
of its X distance (and the planner doesn't know it lost them). As a
result, the first layer of the first tool to use is 5mm off, causing:

* Crash when parking the first tool.
* Layer shift between this initial part and the rest by exactly 5mm in X
  direction.

This only happens if phase stepping is enabled.

Disabling phase stepping for the deretraction (and the following
planner.synchronize()) helps for reasons yet unknown.

BFW-7739.

michalrudolf committed:
Make Pre-nozzle cleaning retraction switchable with persistent value in EEPROM

BFW-7621

o-vesely committed:
Fix negative absorbing heat progress after target change

BFW-7600

o-vesely committed:
Add recalculation of progress if target changes

BFW-7600

michalrudolf committed:
Do not report XBE fans are not spinning on failed modBus read

modBus status communication between XBE na XBuddy is not reliable and can fail,
This should not pop up fan is not spinning error as it is now

BFW-7299

CZDanol committed:
G28: Do not offer to calibrate if cannot calibrate

BFW-7731

CZDanol committed:
auto_retract: can_perform_action → ready_to_extrude

Rename the function to better match what it does (and to avoid confusion with can_retract)

BFW-7721

CZDanol committed:
auto_retract: maybe_retract_from_nozzle: Remove planner shenanigans

They are not necessary there, maybe_retract is NOT interrupting any moves.

BFW-7721

CZDanol committed:
auto_retract: Fix setting wrong e position after deretract

BFW-7721

CZDanol committed:
auto_retract: maybe_deretract before ramming

Just to be sure

BFW-7721

CZDanol committed:
Refactor NozzleCleaningFailedWizard

I'm sorry :(

BFW-7721

CZDanol committed:
Rename nozzle_cleaning_phases → nozzle_cleaning_failed_phases

BFW-7721

CZDanol committed:
NozzleCleaningFailedWizard: Offer to enable auto-retract

BFW-7221

CZDanol committed:
marlin_server: Decouple FSM responses from Requests

Sending responses was previously waiting for the marlin_server confirmation, which was needlessly blocking the calling thread.

Also, it meant that we couldn't blockingly wait for the responses within marlin_server::_server_print_loop, which proved fatal in BFW-7767.

BFW-7767

CZDanol committed:
marlin_server: Discard potentially stray responses

If a response is not consumed, it can stay in server_side_encoded_fsm_response forever.

This means that theoretically, we could have pre-stored a response for some phase hours in advance and it could have been processed when a FSM first popped up.

This commit prevents this by clearing the response with every FSM phase change.

Warnings needed a bit of extra care, because the warning shown is not only determined by the FSM phase, but also by the data.

BFW-7767

CZDanol committed:
marlin_server: Shuffle requests around

- Rename server_queue to request_queue, because it makes more sense
- Put request_queue next to request_flags, because they are closely related
- Add some comments
- Fix include paths

BFW-7767

CZDanol committed:
loadcell: Fix reporting garbage right after  taring

Right after taring. loadcell was reporting nonsense tared_z_load (the offset was not yet picked up).
Loadcell selftest picked the garbage up and failed immediately.

BFW-7791

CZDanol committed:
Introduce MutexAtomic

BFW-7767

michalrudolf committed:
Nozzle cleaning failed wizard: report Attention state to Connect

BFW-7766

CZDanol committed:
ScreenHandler: Do not misuse close for closing dialogs

This was making Close<SomeScreen>() close a dialog above it instead in some cases.

BFW-7718

michalrudolf committed:
Nozzle Cleaning Failed Wizard: Remove unused string

BFW-7766

CZDanol committed:
G28: Fix Z_SAFE_HOMING requirements raise

The AxisHomingRequirement for Z_SAFE_HOMING was accidentally setting initial_level instead of required_level, which was causing the condition to have no effect and homing failures down the line.

Move initial_level as the first member of AxisHomingRequirement so that the mistake has less chance of repeating in the future.

BFW-7792

CZDanol committed:
selftest_axis_z: FIx redscreen at the end

enable_z_probe was enabled for too long (up to the very end).
When the test was saving the results to the EEPROM, the loadcell data starved out (because of the I2C interrupt) and threw a homing fail redscreen.

BFW-7792

michalrudolf committed:
Standard frames: Add error code constructiors for FramePrompt and FrameQRPrompt

BFW-7766

CZDanol committed:
marlin_server: Fix retracting when auto-retracted

This was triggering an auto-retract at the end of a print that had no extruder moves.

BFW-7739

michalrudolf committed:
Replace FrameQRPrompt's QR window with QRErrorUrlWindow

BFW-7766

schrodlm committed:
fixup! Fix first layer selftest

michalrudolf committed:
Replace FrameQRPrompt's link with TextErrorUrl

BFW-7766

michalrudolf committed:
Move and rename Nozzle Cleaning Retraction menu item

BFW-7797

CZDanol committed:
marlin_server: Rework get_last_knob_move

The previous implementation was not a good design, it was unpredicably consuming and unnecesarily complicated.

This one is still not ideal as it disregards multiple moves in one step, but at least it lays ground for fixing that as well.

BFW-7813

michalrudolf committed:
Report nozzle cleaning failed phases to connect

BFW-7766

michalrudolf committed:
Add FrameProgressPrompt constructor with error_code mapper

BFW-7766

CZDanol committed:
Introduce utils::Timer

BFW-7813

CZDanol committed:
temperature: Assume TEMP_RESIDENCY_TIME

I need to a bit of refactoring here and things will be much easier if I don't have to account for this guy not being defined.

It is defined for all our printers.

BFW-7813

CZDanol committed:
wait_for_hotend: Fix fan control

BFW-7813

CZDanol committed:
temperature: Inline isCoolingHotend

And isHeatingHotend is not used anywhere anymore.

BFW-7813

CZDanol committed:
Remove dead M701_M702.cpp

We have our stub in marlin_stubs

BFW-7813

CZDanol committed:
temperature: Unify reaching temperatures checks

BFW-7813

CZDanol committed:
Implement new SafetyTimer

Integration will come in the next commit.

BFW-7813

CZDanol committed:
Remove old safety_timer API

BFW-7813

CZDanol committed:
Partially migrate to the new SafetyTimer

Still a few more things to do, will be done in the following commits

BFW-7813

CZDanol committed:
Remove old SafetyTimer

BFW-7813

CZDanol committed:
marlin_server: Remove knob tracking

It was only used by the old SafetyTimer.

BFW-7813

michalrudolf committed:
git subrepo pull --force lib/Prusa-Error-Codes

subrepo:
  subdir:   "lib/Prusa-Error-Codes"
  merged:   "3af5322e89"
upstream:
  origin:   "git@github.com:prusa3d/Prusa-Error-Codes.git"
  branch:   "master"
  commit:   "3af5322e89"
git-subrepo:
  version:  "0.4.9"
  origin:   "https://github.com/ingydotnet/git-subrepo"
  commit:   "cce3d93"

CZDanol committed:
Migrate pause to the new SafetyTimer

IPause and things around were there only for the old SafetyTimer.
We have a new approach now.

BFW-7813

CZDanol committed:
Remove another safety timer

Yah, why not have two?

BFW-7813

CZDanol committed:
!BOARD_IS_DWARF -> BOARD_IS_MASTER_BOARD

BFW-7813

CZDanol committed:
temperature: Move setTargetXX to cpp

BFW-7813

CZDanol committed:
power_panic: Use are_all_temperatures_reached instead of enqueing M gcodes

BFW-7813

CZDanol committed:
power_panic: Use std::array for nozzle temps

This allows copying by value.

BFW-7813

CZDanol committed:
SafetyTimer: Disable during heater selftest

BFW-7813

CZDanol committed:
SafetyTimer: Improve power panic support

BFW-7813

CZDanol committed:
marlin_server: Fix missing return in resuming_reheating

BFW-7813

CZDanol committed:
marlin_server: Remove unpause_nozzle bullshido

Our SafetyTimer now automatically resumes the temperatures, we don't need this.

BFW-7813

CZDanol committed:
marlin_set_variable: Do not alter resume.nozzle_temp

This one is just so weird. You can change a hotend temperature in many other ways and those would keep the resume temperature the same.

Also, if you set the target temp to 0 during pause, this would make the hotend return to 0 after unpausing.

BFW-7813

CZDanol committed:
ScreenFSM: Normalize CaptureNormalWindow(inner_frame)

BFW-7813

CZDanol committed:
ScreenPreheat: Drop duplicit header

A header is already provbided by ScreenFSM

BFW-7813

CZDanol committed:
DialogLoadUnload: Separate DialogFSM

BFW-7813

CZDanol committed:
DialogHandler: Reduce mem_space

We don't have that big dialogs anymore.

BFW-7813

CZDanol committed:
Fix ClientFSM::BeltTuning being used for both HAS_BELT_TUNING and HAS_MANUAL_BELT_TUNING

BFW-7813

CZDanol committed:
SafetyTimer: Add GUI

Add some gui for reheating, visualising the progress and offering to abort.

BFW-7813

michalrudolf committed:
planner: Always recalculate segment length

By merging short segments producing few steps, the effective segment
length of the current block might be different compared to what was
hinted.

Even if the block is marked as no-discard, there's no telling if any
number of previous segments was discarded, thus changing the effective
segment length and calculating an incorrect acceleration profile.

Also, always use XYZ axes for the segment length, even if the length is
short: this fixes a bug where a short no-discard segment could
incorrectly use a zero-length extrusion as the total distance, where in
fact the move was on any other axis.

Cherry-pick from Yuri's refactoring
BFW-7827

michalrudolf committed:
planner: Skip velocity calculations on zero-length segments

Also reset all velocities to zero otherwise, to ensure such blocks are
not used in other planning stages.

Cherry-pick from Yuri's refactoring

BFW-7827

michalrudolf committed:
phstep: Correctly process zero-interval moves

Cherry-pick from Yuri's refactoring
BFW-7827

michalrudolf committed:
phstep: Use the move end time of the last planned move
When sorting for the next pending event, we can refer directly to the
last planned target of a phase stepped axis, which increases the
look-ahead of the step generators to the size allowed by the target
queue.

By saving the value at the time we enqueue the we avoid having to store
it in each move, the need to reduce its precision, while also avoiding
using the current (and possibly invalid) is_state

Cherry-pick from Yuri's refactoring
BFW-7827

CZDanol committed:
MK4/XL: enable Z_NOZZLE_PARK_POINT_MIN

BFW-7811

CZDanol committed:
Rework stepper timeout

BFW-7383

CZDanol committed:
Drop M85

It was quite shaggily implemented,
if anyone would have actually used it,
he would probably get himself into trouble.

Also, it was leeching on gcode.reset_stepper_timeout() and messing things together with the stepper inactivity timeout.

BFW-7383

CZDanol committed:
Drop old stepper timeout functionality

BFW-7383

CZDanol committed:
Drop unused EXTRUDER_RUNOUT_PREVENT

BFW-7383

CZDanol committed:
Remove now useless reset_stepper_timeout

BFW-7383

CZDanol committed:
marlin_server: Fix is_processing not considering selftests

BFW-7383

vorner committed:
phase stepping: Steal next target if starved

If the phase stepping gets starved, allow it to steal the next_target -
a move that's staged for possible small adjustment of end position.

In general, this situation shouldn't happen, but for some reason, under
very specific circumstances the next move to flush this one wasn't
coming, leading to serious layer shifts on XL.

Implementing this fallback mechanism as a robustness measure.

BFW-7739.

vorner committed:
XL: Disable autoretract workaround

BFW-7739.

CZDanol committed:
Marlin: remove manage_inactivity() ignore_stepper_queue parameter

BFW-7383

CZDanol committed:
idle(): Drop no_stepper_sleep parameter

BFW-7383

CZDanol committed:
Introduce EnumBitset

BFW-7813

CZDanol committed:
Cherry-pick fixes

BFW-7813

CZDanol committed:
selftest: Unify what being completed means

BFW-7095

CZDanol committed:
marlin_server: FIx is_processing not building on minimal

BFW-7383

michalrudolf committed:
GUI: Increase Fan selftest info's row count to fit translations

BFW-7844

michalrudolf committed:
GUI: Change fan selftest items' (except info) font small to fit translated texts

BFW-7844

CZDanol committed:
FramePrompt: Fix improper CaptureNormalWindow

Caused by cherry-picking BFW-7766 without cherry-picking BFW-7702, the resulting code got broken.

BFW-7846

CZDanol committed:
SafetyTimer: Fix excessive FSM changes

BFW-7851

CZDanol committed:
Pause: Simplify AutoRestore guard

BFW-7851

CZDanol committed:
Add SafetyTimerNonBlockingGuard

We need a way to mark some actions/movements as safe to be done even during SafetyTimer resuming (for example filament change)

BFW-7851

CZDanol committed:
Pause: Make non-safetytimer blocking

Prevent the "waiting for temperature restore" from triggering - the Pause manages temperature safety for extrusion internally

BFW-7851

CZDanol committed:
Fix steppers disabling during print pause

BFW-7852

CZDanol committed:
Fix print pause-SafetyTimer clash

If SafetyTimer was active when a print pause started, zeroes were stored to the restore temps.

BFW-7852

michalrudolf committed:
Skip preheat during Filament change if nozzle is auto retracted

BFW-7841

CZDanol committed:
marlin_server: Fix print_reheat_ready

Fix a regression introduced by temperature check.

BFW-7852

CZDanol committed:
Get rid of nozzle_temp_paused

It was only causing bugs and directly copying the printer paused state.

BFW-7845

CZDanol committed:
marlin_server: Remove shortcut in resuming_begin

Code flow should reach that point naturally through the state succession. This was only adding space duplicity and opportunities for mistakes.

BFW-7845

CZDanol committed:
marlin_server: Inline and simplify print_reheat_ready

This also fixes soft-lock that could have happened if someone changed to a nonzero target temp during reheating.

BFW-7845

CZDanol committed:
config_store: Implement metrics

BFW-7758

CZDanol committed:
metrics: Sort list by metric name

Allows easier navigation in the GUI metrics list.

BFW-7830

CZDanol committed:
qr.hpp: Fix missing include

BFW-7758

CZDanol committed:
ubl: Fixup after refactorings

Before the refactorings, the `if (std::isnan(measured_z))` was inside the probe_major_points body and not in a lambda passed to for_each_grid_point, so its return returned from the probe_major_points.

Now, this does not happen and STOW_PROBE is called at the end of the function.

It is pribably a good thing that we call restore_ubl_active_state_and_leave now even on failures, but we should at least keep the early return in the sense that on failure, we won't be probing any more points.

BFW-7801

CZDanol committed:
G29: Do not rehome on probing fail

On retry, the G29 only remeasures points that failed, not all of them, so the previously passed points will likely get screwed up by the Z rehome.

Also, even if we reprobed all of these, the fail can happen on an additional probe around the purge area, in which case we would still not reprobe the whole bed.

BFW-7801

CZDanol committed:
Fix unittests failing

CZDanol committed:
Differentiate not done and skipped selftests

BFW-7876

o-vesely committed:
Separate printer setup into hardware and network configuration steps

More granular tracking of the printer's setup progress. And display only the ones missing when printer si selectively factory reset.

BFW-7867

CZDanol committed:
WUI: Pass std::string_view by reference

Trying to save some stack

BFW-7881

CZDanol committed:
GCodeReader: Allow opening post-construction

BFW-7881

CZDanol committed:
wui: Reduce stack usage during preview.get

ConnectionState is a 400B structure.
GCodePreview is a 400B structure.
AnyGcodeFormatReader is a 400B structure.

out.next is of type ConnectionState.

The tcpip stack was overflowing, at which point the accept() function was taking 1560 B of stack.

BFW-7881

vorner committed:
Fix wrong ordering

The acquire ordering needs to be on read operation, so store doesn't cut
it.

BFW-7739.

michalrudolf committed:
Don't execute maybe_retract in print end if mmu is enabled

Unloading the nozzle is handled by Slicer and also frequent filament changes cause
that retract tracker cannot properly hold valid value -> retraction at pre_finalize_print is certain

BFW-7848

CZDanol committed:
SystemInit: Reset all peripherals

We noticed some BSODs caused by HAL_DMA_ERROR_TIMEOUT in hdma_spi6_tx HAL_DMA_Init.

We suspect that the bootloader can be leaving the DMA running when passing control over to the firmware, which then results in the DMA init failure.

To remedy this, we now reset all peripherals to initial state in the initialization procedure.

BFW-7880

danopernis committed:
Update translations

Commit:
Planner: do not fail quick_stop spuriously

It shall always set the stop_pending. Only in case stop_pending was false before, update the timestamp.

Probably the root cause of BFW-7473 and many other quick-stop related, unexplainable and hardly reproducible issues (door switch, stopping filament, pause, power panic etc.)

vorner committed:
Better synchronization Planner -> MoveISR

When the planner finishes recalculation of speeds, notify the MoveISR
that it may take another move(s) from there, they are fresh. No need
to wait up to 1ms.

BFW-7873.

vorner committed:
Critical section around planner recompute

To prevent move ISR starvation. It's hopefully fine, as that one does
only bit of computations and doesn't wait on any peripherals or other
tasks.

BFW-7873.

vorner committed:
Bigger buffers in phase stepping

This prevents starving events in phase stepping in case of high-speed
curved moves, where the input shaper produces insane amounts of very
very tiny micro moves. This in turn prevents layer shifts.

This is probably just a workaround. The input shaper (running in the
stepper interrupt, which is very high priority) should have enough time
to refill the buffers. It seems that once it fills the output buffer, it
doesn't get called again when it empties and instead waits for some
other event on other axis (Z or E). This needs some further
investigation, but for now this seems to improve the situation
significantly.

The new value of 128 was discovered by experimental means with a very
hard-core layer shift reproducer, so it should be safe for any practical
means.

BFW-7873.

michalrudolf committed:
Migrate whole distance of standard ramming sequence to new auto_retract length system

BFW-7776

danopernis committed:
Add reusable 'Happy printing!' message box

BFW-7945

danopernis committed:
Show 'Happy printing!' only once

BFW-7945

danopernis committed:
Show 'Happy printing!' even without selftests running

BFW-7945

vorner committed:
Wake up the move ISR when running low

When running low on data in precise stepper or phase stepping, try to
wake up the move interrupt early, to refill.

* We hope there are some move segments cached to be processed even in
  case planner is currently blocking more of them.
* We hope there will be enough time or that the stall will be as short
  as possible if there isn't; certainly better situation than waiting
  for up to a whole 1ms.
* Remove the critical section from the planner side again - even if the
  moves are blocked, we may be able to do the refill from the cached
  data. Remove it with just an atomic flag so the move ISR will just
  skip looking there while it's being updated.

BFW-7936, BFW-7873.

vorner committed:
Make the MoveISR greedy

Instead of limiting the amount of segments processed to at most 1 per
1ms (or one call now), process as much as possible. We've observed
segments with shorter times than 1ms.

Even when we have the wakeup on running dry, it's better to keep the
buffers full and have smaller amount of the interrupt calls for
performance.

BFW-7873.

vorner committed:
Adjust phase stepping queue sizes

With the running-dry wakeup, we can lower the queues again a bit - we
don't necessarily have to survive for a whole 1ms in that case.

BFW-7936, BFW-7873.

danopernis committed:
Fix BSOD on M501

BFW-7942

vorner committed:
MoveISR: Track progress of consuming moves

Instead of guessing if taking moves out of planner made progress
indirectly, just return it from the function. This is a slight
optimization of the move_isr loop.

BFW-7873.

CZDanol committed:
trinamic: Properly condition phase_stepping usage

BFW-7942

CZDanol committed:
reset_trinamic_drivers: Disable phstep only if initialized

EnsureDisabled was triggering phstep::assert_initialized, because the reset_trinamic_drivers was being called on printer boot before the phstep is initialized.

BFW-7942

CZDanol committed:
marlin_server: media recovery vs unpausing

Differentiate between those two situations.
We're doing this to allow showing warnings on unpause failure, but only if it's not an automatic background attempt.

BFW-7938

CZDanol committed:
marlin_server: Separate out prefetch_status_to_warning

I will need to use it on another place as well in the following commit.

BFW-7938

CZDanol committed:
marlin_server: Show media errors on unpausing

Without this, the "resume" button does nothing if there is an error. Or you get a flash of "resuming" and then you go back.

BFW-7938

CZDanol committed:
marlin_server: No prefetch_start in resuming_begin

Media prefetch was already stared in the BufferData state that is preceding resuming begin. Calling start here only results in the data being discarded and having to be refetched again.

Problem introduced in BFW-6594

BFW-7938

michalrudolf committed:
Nozzle Cleaning Wizard: Set cooldown temp before user confirmation

To avoid oozing with high temps while waiting for user

BFW-7802

CZDanol committed:
Post-cherry fixups

BFW-7938

CZDanol committed:
emergency_stop: Don't BSOD during probing

Probing/homing moves have a bit of a special place, because the endstops are enabled during them.
Endstops, when triggered, cause quickstops.

So during homing, we don't want to do the "standard" segment-and-maybe-block thing, because weird things would start happening.

Instead of that (and instead of just BSODing), we now do a quick_stop (which is the same as the endstop being triggered) and block after we get out of the probing code.

BFW-7544

CZDanol committed:
probe: Segment lift moves

If we don't segment them, emergency_stop cannot interrupt it, and we end up with a BSOD.

BFW-7544

vorner committed:
M18: Fix 0=disabled case

The M18 called with 0 was meant to set the stepper timeout do
"disabled", not to set interval to 0 (which had weird side effects, like
the printer not moving in dock calibration).

BFW-7993.

vorner committed:
Bigger ISR stack size

Recent changes made it overflow from time to time (they either made the
stack of some interrupt function larger or the chance of multiple
interrupt stacking on each other got higher).

Trying to add 100 bytes (+alignment), as we don't have a direct way to
measure how much we need.

BFW-7873.

3d-gussner commented:
@me25542 Thanks for the PR. Added a comment in the internal ticket.
As it uses quite a bit of extra resources it may become an issue on MINI.

me25542 commented:
Thanks.
You're right that this might not work on the MINI; resource usage is increased somewhat. However, it *does* now seem to build OK for the MINI, after some optimizations (it didn't build for the MINI when this PR was first opened), so I don't know for sure.

michalrudolf committed:
Unify HAS_AUTOMATIC_CHAMBER_VENTS and HAS_MANUAL_CHAMBER_VENTS

BFW-8021

michalrudolf committed:
Add auto/manual switching of vent mechanism

BFW-8021

michalrudolf committed:
Add menu item for vent mechanism control

BFW-8021

michalrudolf committed:
Add auto vent control for CoreOne

BFW-8021

michalrudolf committed:
Avoid vent mechanism during homing moves on CoreOne

BFW-8021

michalrudolf committed:
Change persistent variable check_chamber_vent_state to ItemFlag::features

BFW-8021

michalrudolf committed:
Set chamber vent state after M870 is called

BFW-8021

michalrudolf committed:
Avoid double pop up of Happy Printing

BFW-7893

CZDanol committed:
crash_recovery: Use print_resume

Crash recovery was hard-switching to Resuming_Begin print state, which:
- Did not trigger media_prefetch_call()
- Skipped the resuming buffering phase.

Using print_resume unifies the functionality and print_resume makes sure that all necessary steps are done.

BFW-8082

me25542 commented:
Extended testing revealed that optimizations made to enable this feature on the MINI broke things, so I have switched to working code, and disabled this feature on the MINI.

me25542 committed:
gui: Add option to hide file extensions

BFW-6704

me25542 commented:
I accidentally closed this somehow when updating my branch.

bkerler commented:
@me25542 Does this still break things on MINI ?

me25542 commented:
This still doesn't work on MINI; It does compile for everything else and seems to work on my MK4S.
I added guards so building the MINI firmware still works, but doesn't have my changes.

me25542 commented:
After looking into it more, it seems like I won't be able to get this working on the MINI, and other printers are stuck with increased memory usage (up to around 728 extra bytes).
